File: completed/1498.java
import java.util.*;

class Solution {
    private static final int M = (int) 1e9 + 7;

    public int numSubseq(int[] nums, int target) {
        Arrays.sort(nums);

        int[] modularPow2 = new int[nums.length];
        modularPow2[0] = 1;
        for (int i = 1; i < nums.length; i++) {
            modularPow2[i] = (modularPow2[i - 1] << 1) % M;
        }

        int count = 0;
        int left = 0;
        int right = nums.length - 1;
        while (left <= right) {
            // left is min, right is max since nums is sorted
            if (nums[left] + nums[right] <= target) {
                // left and right must be picked,
                // so the (right - left) elements
                // in between can either be picked or not picked
                count += modularPow2[right - left];
                count %= M;
                left++;
            } else {
                right--;
            }
        }
        return count;
    }
}

File: completed/1498.cpp
class Solution {
private:
    int M = 1e9 + 7;

    int exp(long base, long exponent) {
        long result = 1;
        while (exponent > 0) {
            if (exponent & 1) {
                result *= base;
                result %= M;
            }
            exponent >>= 1;
            base *= base;
            base %= M;
        }
        return (int) result;
    }

public:
    int numSubseq(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());

        long total = 0;
        for (int i = 0; i < nums.size(); i++) {
            int complement = target - nums[i];
            int left = i;
            int right = nums.size() - 1;
            while (left <= right) {
                int mid = (left + right) / 2;
                if (nums[mid] <= complement) {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            }

            if (left <= i) {
                continue;
            }

            int index = left - i - 1;
            int count = exp(2, index);
            total = (total + count) % M;
        }
        return total;
    }
};
